src/nitbuilder.nit

```diff
@@ -0,0 +1,286 @@
+# This file is part of NIT ( http://www.nitlanguage.org ).
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Instantiation and transformation of language entities in AST and model.
+# The module provides two services:
+#
+# - Nitbuilder : Allows according to the context (i.e. if we already have the representation of the model, or just the name)
+#				to create entities of the model/ast and to associate it together. Nitbuilder also provides a callsite
+#				creation service using the current infrastructure (TypeVisitor).
+#
+# - Addition of a `create_ast_representation` method in` MEntity` and `MPropDef`, this method simplifies
+#	the creation of the ast based on the information available in the model. It is easier to create model entities than ast.
+#
+module nitbuilder
+
+import modelbuilder
+intrude import astbuilder
+intrude import modelize_property
+intrude import scope
+intrude import typing
+
+# Provides the construction of model entities and their associated ast.
+class NitBuilder
+
+	# The toolcontext instance
+	var toolcontext: ToolContext
+
+	# Instance of the AST constructor
+	var ast_builder: ASTBuilder
+
+	# Try to get MMethod property if exist in the given mclassdef. return new `MMethod` if not exist.
+	private fun get_mmethod(name: String, mclassdef: MClassDef, visibility: nullable MVisibility): MMethod do
+		if visibility == null then visibility = public_visibility
+		var mproperty = toolcontext.modelbuilder.try_get_mproperty_by_name(null, mclassdef, name).as(nullable MMethod)
+		if mproperty == null then mproperty = new MMethod(mclassdef, name, mclassdef.location, visibility)
+		return mproperty
+	end
+
+	# Creation of a new method (AST and model representation) with the given name.
+	# See `create_method_from_property` for more information.
+	fun create_method_from_name(name: String, mclassdef: MClassDef, msignature: nullable MSignature, visibility: nullable MVisibility): AMethPropdef do
+		var mproperty = get_mmethod(name, mclassdef, visibility)
+		return create_method_from_property(mproperty, mclassdef, msignature)
+	end
+
+	# Same as `create_method_from_name` except the returned AMethPropdef is abstract.
+	# See `create_abstract_method_from_property` for more information.
+	fun create_abstract_method_from_name(name: String, mclassdef: MClassDef, msignature: nullable MSignature, visibility: nullable MVisibility): AMethPropdef do
```

Comment:
The API will be nicer if you merged this one with the previous and used a `is_abstract` parameter.